crates/nsis-process/src/lib.rs

```diff
@@ -118,97 +122,212 @@ fn KillProcessCurrentUser() -> Result<(), Error> {
-            &mut info_length as *mut u32,
-        );
-
-        // GetTokenInformation always returns 0 for the first call so we check if it still gave us the buffer length
-        if info_length == 0 {
-            return sid;
-        }
-
-        let info = vec![0u8; info_length as usize].as_mut_ptr() as *mut TOKEN_USER;
-
-        if GetTokenInformation(
-            token_handle,
-            TokenUser,
-            info as *mut c_void,
-            info_length,
-            &mut info_length,
-        ) == 0
-        {
-            return sid;
-        }
-
-        sid = Some((*info).User.Sid)
+    let handle = OwnedHandle::new(OpenProcess(PROCESS_QUERY_INFORMATION, 0, pid));
+    if handle.is_invalid() {
+        return None;
+    }
+    let mut token_handle = OwnedHandle::new(HANDLE::default());
+    if OpenProcessToken(*handle, TOKEN_QUERY, &mut *token_handle) == FALSE {
+        return None;
     }
 
-    CloseHandle(token_handle);
-    CloseHandle(handle);
+    let mut info_length = 0;
+    GetTokenInformation(
+        *token_handle,
+        TokenUser,
+        ptr::null_mut(),
+        0,
+        &mut info_length,
+    );
+    // GetTokenInformation always returns 0 for the first call so we check if it still gave us the buffer length
+    if info_length == 0 {
+        return None;
+    }
 
-    sid
+    let mut buffer = vec![0u8; info_length as usize];
+    let info = buffer.as_mut_ptr() as *mut TOKEN_USER;
+    if GetTokenInformation(
+        *token_handle,
+        TokenUser,
+        info as *mut c_void,
+        info_length,
+        &mut info_length,
+    ) == FALSE
+    {
+        None
+    } else {
+        Some((*info).User.Sid)
+    }
 }
 
 fn get_processes(name: &str) -> Vec<u32> {
     let current_pid = unsafe { GetCurrentProcessId() };
     let mut processes = Vec::new();
 
     unsafe {
-        let handle = CreateToolhelp32Snapshot(TH32CS_SNAPPROCESS, 0);
+        let handle = OwnedHandle::new(CreateToolhelp32Snapshot(TH32CS_SNAPPROCESS, 0));
 
         let mut process = PROCESSENTRY32W {
             dwSize: mem::size_of::<PROCESSENTRY32W>() as u32,
             ..mem::zeroed()
         };
 
-        if Process32FirstW(handle, &mut process) != 0 {
-            while Process32NextW(handle, &mut process) != 0 {
+        if Process32FirstW(*handle, &mut process) == TRUE {
+            while Process32NextW(*handle, &mut process) == TRUE {
                 if current_pid != process.th32ProcessID
                     && decode_utf16_lossy(&process.szExeFile).to_lowercase() == name.to_lowercase()
                 {
                     processes.push(process.th32ProcessID);
                 }
             }
         }
-
-        CloseHandle(handle);
     }
 
     processes
 }
 
+/// Return true if success
+///
+/// Ported from https://devblogs.microsoft.com/oldnewthing/20190425-00/?p=102443
+unsafe fn run_as_user(command: &str, arguments: &str) -> bool {
+    let hwnd = OwnedHandle::new(GetShellWindow());
```

Comment:
we shouldn't use OwnedHandle here since this is a HWND not HANDLE and so it shouldn't call CloseHandle on drop